File: skills/medtrum-reddit-spider/scripts/reddit_search.mjs
#!/usr/bin/env node
/**
 * Reddit search runner via browser automation.
 * Used when direct HTTP requests are blocked (403/anti-bot).
 */

import { ensureHostBrowserBridge } from '../../web-access/scripts/host-bridge.mjs';

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function proxyRequest(endpoint, options = {}) {
  const url = `http://127.0.0.1:3456${endpoint}`;
  const response = await fetch(url, {
    ...options,
    signal: AbortSignal.timeout(60000),
  });
  const text = await response.text();
  try { return JSON.parse(text); } catch { return text; }
}

async function createTarget(url) {
  return proxyRequest(`/new?url=${encodeURIComponent(url)}`);
}

async function evalScript(targetId, expression) {
  return proxyRequest(`/eval?target=${encodeURIComponent(targetId)}`, {
    method: 'POST',
    headers: { 'Content-Type': 'text/plain; charset=utf-8' },
    body: expression,
  });
}

async function closeTarget(targetId) {
  return proxyRequest(`/close?target=${encodeURIComponent(targetId)}`);
}

// Extract Reddit search results from DOM (for Reddit's SDUI structure)
const EXTRACT_POSTS_SCRIPT = `
(function() {
  const results = [];
  const containers = document.querySelectorAll('[data-testid="search-sdui-post"], [data-testid="search-post-unit"], [data-testid="sdui-post-unit"]');
  containers.forEach(container => {
    try {
      const titleEl = container.querySelector('[data-testid="post-title-text"]') || container.querySelector('a[href*="/comments/"]');
      const title = titleEl ? titleEl.textContent.trim() : '';
      const linkEl = container.querySelector('a[href*="/comments/"]');
      const href = linkEl ? linkEl.getAttribute('href') : '';
      const permalink = href ? (href.startsWith('/') ? 'https://www.reddit.com' + href.split('?')[0] : href.split('?')[0]) : '';
      if (!permalink || !permalink.includes('/comments/')) return;
      const subMatch = permalink.match(/reddit\\.com\\/r\\/([^\\/]+)/);
      const subreddit = subMatch ? 'r/' + subMatch[1] : 'r/unknown';
      const authorEl = container.querySelector('a[href^="/user/"]');
      let author = authorEl ? authorEl.getAttribute('href').replace('/user/', '').replace('/comments/', '') : '[unknown]';
      const timeEl = container.querySelector('time');
      const datetime = timeEl ? timeEl.getAttribute('datetime') : '';
      const timeText = timeEl ? timeEl.textContent.trim() : '';
      results.push({ datetime, timeText, subreddit, author, title, selftext: '', score: 0, numComments: 0, permalink });
    } catch (e) {}
  });
  // Fallback: scan all comment links
  if (results.length === 0) {
    const allLinks = document.querySelectorAll('a[href*="/comments/"]');
    allLinks.forEach(link => {
      const href = link.getAttribute('href');
      const permalink = href.startsWith('/') ? 'https://www.reddit.com' + href.split('?')[0] : href.split('?')[0];
      if (permalink.includes('/comments/') && !results.find(r => r.permalink === permalink)) {
        const subMatch = permalink.match(/reddit\\.com\\/r\\/([^\\/]+)/);
        results.push({
          datetime: '', timeText: '', subreddit: subMatch ? 'r/' + subMatch[1] : 'r/unknown',
          author: '[unknown]', title: link.textContent.trim().substring(0, 100), selftext: '', score: 0, numComments: 0, permalink
        });
      }
    });
  }
  return JSON.stringify(results);
})();
`;

const SCROLL_SCRIPT = `(function(){
  for (let i = 0; i < 5; i++) {
    window.scrollBy(0, 80);
  }
})();`;

function parseDatetime(datetime, timeText) {
  if (datetime) return datetime;
  if (timeText) {
    const now = new Date();
    const match = timeText.match(/(\d+)\s+(hour|day|week|month|year)s?\s+ago/i);
    if (match) {
      const value = parseInt(match[1], 10);
      const unit = match[2].toLowerCase();
      switch (unit) {
        case 'hour': now.setHours(now.getHours() - value); break;
        case 'day': now.setDate(now.getDate() - value); break;
        case 'week': now.setDate(now.getDate() - value * 7); break;
        case 'month': now.setMonth(now.getMonth() - value); break;
        case 'year': now.setFullYear(now.getFullYear() - value); break;
      }
      return now.toISOString();
    }
  }
  return null;
}

async function main() {
  const keyword = process.argv[2] || '';
  const days = parseInt(process.argv[3], 10) || 30;
  const maxScrolls = parseInt(process.argv[4], 10) || 8;

  if (!keyword) {
    console.error('错误：关键词不能为空');
    process.exit(1);
  }

  const url = `https://www.reddit.com/search/?q=${encodeURIComponent(keyword)}&sort=new&type=link`;

  await ensureHostBrowserBridge();

  console.error('[reddit] Opening...');
  let targetId = null;
  try {
    const created = await createTarget(url);
    targetId = created?.targetId || created?.target?.id;
  } catch (e) {
    console.error('无法创建浏览器 target:', e.message);
    process.exit(1);
  }

  console.error('Target ID:', targetId);
  await sleep(5000);

  // Preflight: verify page is healthy
  const preflightRaw = await evalScript(targetId, `(function(){
    var title = document.title || '';
    if (title.includes('blocked') || title.includes('429')) return JSON.stringify({ ok: false, error: 'BLOCKED', detail: 'Reddit rate-limited' });
    if (!title) return JSON.stringify({ ok: false, error: 'EMPTY_PAGE', detail: 'Page did not load' });
    var posts = document.querySelectorAll('[data-testid="search-sdui-post"], [data-testid="search-post-unit"], [data-testid="sdui-post-unit"]');
    if (posts.length === 0) return JSON.stringify({ ok: false, error: 'NO_RESULTS', detail: 'No search result units found' });
    return JSON.stringify({ ok: true, title, postUnits: posts.length });
  })()`);
  let preflight = { ok: false, error: 'unknown' };
  try { preflight = JSON.parse(preflightRaw); } catch(e) {}
  if (!preflight.ok) {
    console.error('REDDIT_PREFLIGHT_FAILED:' + (preflight.error || 'unknown') + ' ' + (preflight.detail || ''));
    process.exit(2);
  }

  let allPosts = [];
  const seenUrls = new Set();
  let noNewRounds = 0;
  const noNewLimit = 3;

  // Extract posts with smooth scrolling to trigger Reddit's infinite scroll
  for (let i = 0; i < maxScrolls; i++) {
    console.error(`[reddit] round ${i+1}...`);
    try {
      const rawResult = await evalScript(targetId, EXTRACT_POSTS_SCRIPT);
      if (rawResult) {
        try {
          const posts = JSON.parse(rawResult);
          let newCount = 0;
          for (const post of posts) {
            if (post.permalink && !seenUrls.has(post.permalink)) {
              seenUrls.add(post.permalink);
              allPosts.push({ ...post, datetime: parseDatetime(post.datetime, post.timeText) });
              newCount++;
            }
          }
          if (newCount === 0) {
            noNewRounds++;
          } else {
            noNewRounds = 0;
          }
        } catch (e) { /* JSON parse fail, continue */ }
      }
    } catch (e) {
      console.error(`[reddit] extract error: ${e.message}`);
    }
    console.error(`[reddit] ${i + 1}: ${allPosts.length} posts`);

    if (noNewRounds >= noNewLimit) {
      console.error(`[reddit] no new, stopped`);
      break;
    }

    if (i < maxScrolls - 1) {
      // Smooth scroll using pre-defined SCROLL_SCRIPT (function-wrapped)
      try {
        await evalScript(targetId, SCROLL_SCRIPT);
        await sleep(2500);
      } catch (e) {
        // Fallback: try single scrollBy
        try {
          await evalScript(targetId, `(function(){ window.scrollBy(0, 400); })()`);
          await sleep(2500);
        } catch (e2) {
          console.error('[reddit] scroll failed, stopping');
          break;
        }
      }
    }
  }

  // Filter by days
  const cutoffDate = new Date();
  cutoffDate.setDate(cutoffDate.getDate() - days);
  const filteredPosts = allPosts.filter(post => {
    if (!post.datetime) return true;
    return new Date(post.datetime) >= cutoffDate;
  });

  const validItems = [];
  let dropped = 0;
  for (const post of filteredPosts) {
    const date = post.datetime ? new Date(post.datetime).toISOString().slice(0, 10) : '';
    const author = String(post.author || '').trim();
    const content = String((post.title || '') + (post.selftext ? ' - ' + post.selftext : '')).trim();
    const url = String(post.permalink || '').trim();

    if (!date) { dropped++; continue; }
    if (!content || content.length < 10) { dropped++; continue; }
    if (url && !url.startsWith('https://www.reddit.com/')) { dropped++; continue; }

    validItems.push({ date, author: author || 'unknown', content: content.slice(0, 300), url });
  }

  const result = {
    platform: 'Reddit',
    keyword,
    retrievedAt: new Date().toISOString(),
    queryUrl: url,
    preflight,
    scrollStatus: { allCollected: allPosts.length, filtered: filteredPosts.length },
    total: validItems.length,
    dropped,
    items: validItems,
  };
  console.log(JSON.stringify(result, null, 2));

  try { await closeTarget(targetId); } catch (e) {}
}

main().catch(err => {
  console.error('执行失败:', err.message);
  process.exit(1);
});

File: skills/medtrum-pcm/scripts/reddit_search.py
#!/usr/bin/env python3
"""
Reddit search wrapper — 双通道降级策略。

通道 1（优先）：调用 reddit-data skill 的 reddit.py search（HTTP JSON API）
通道 2（降级）：HTTP API 被反爬封禁时，切回原浏览器方案 reddit_search.mjs

用法: python3 reddit_search.py <keyword> [days]
"""

import json
import os
import subprocess
import sys
from datetime import datetime, timezone, timedelta

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
REDDIT_PY = "/app/runtime/skills-user/reddit-data/scripts/reddit.py"
REDDIT_MJS = os.path.join(SCRIPT_DIR, "reddit_search.mjs")


def run_api(keyword, days):
    """通道 1：HTTP JSON API"""
    cmd = [
        sys.executable, REDDIT_PY,
        "search", keyword,
        "--sort", "new",
        "--time", "month",
        "--limit", "100",
    ]
    proc = subprocess.run(cmd, capture_output=True, text=True, timeout=120)

    if proc.returncode != 0:
        return None, proc.stderr

    try:
        raw = json.loads(proc.stdout)
    except json.JSONDecodeError:
        return None, proc.stdout

    # API 返回了 error 字段
    if "error" in raw:
        return None, json.dumps(raw)

    posts = raw.get("posts", [])
    cutoff = datetime.now(timezone.utc) - timedelta(days=days)

    valid_items = []
    dropped = 0

    for post in posts:
        created_str = post.get("created_utc", "")
        if not created_str:
            dropped += 1
            continue
        try:
            if " UTC" in created_str:
                created_dt = datetime.strptime(created_str, "%Y-%m-%d %H:%M:%S UTC").replace(tzinfo=timezone.utc)
            else:
                created_dt = datetime.fromisoformat(created_str)
        except (ValueError, TypeError):
            dropped += 1
            continue
        if created_dt < cutoff:
            dropped += 1
            continue

        date_str = created_dt.strftime("%Y-%m-%d")
        author = post.get("author", "unknown").strip() or "unknown"
        title = post.get("title", "") or ""
        selftext = post.get("selftext", "") or ""
        content = (title + " - " + selftext if selftext else title)[:300].strip()
        permalink = post.get("permalink", "") or ""
        url = f"https://www.reddit.com{permalink}" if permalink.startswith("/") else permalink

        if not content or len(content) < 10:
            dropped += 1
            continue

        valid_items.append({
            "date": date_str,
            "author": author,
            "content": content,
            "url": url,
        })

    result = {
        "platform": "Reddit",
        "keyword": keyword,
        "retrievedAt": datetime.now(timezone.utc).isoformat(),
        "queryUrl": f"https://www.reddit.com/search/?q={keyword}&sort=new",
        "total": len(valid_items),
        "dropped": dropped,
        "items": valid_items,
    }
    return result, None


def run_browser(keyword, days):
    """通道 2：浏览器降级（原 reddit_search.mjs）"""
    cmd = [
        "node", REDDIT_MJS,
        keyword, str(days),
    ]
    proc = subprocess.run(cmd, capture_output=True, text=True, timeout=180)

    if proc.stderr:
        print(proc.stderr, file=sys.stderr)

    if proc.returncode != 0:
        if "REDDIT_PREFLIGHT_FAILED" in (proc.stderr or ""):
            print("PREFLIGHT_FAILED: browser channel preflight failed", file=sys.stderr)
            sys.exit(2)
        sys.exit(proc.returncode)

    try:
        return json.loads(proc.stdout)
    except json.JSONDecodeError:
        print("Error: browser channel returned invalid JSON", file=sys.stderr)
        sys.exit(1)


def main():
    keyword = sys.argv[1] if len(sys.argv) > 1 else ""
    days = int(sys.argv[2]) if len(sys.argv) > 2 else 30

    if not keyword:
        print("Error: keyword required", file=sys.stderr)
        sys.exit(1)

    # ---- 通道 1：优先尝试 HTTP API ----
    print(f"[reddit] 通道1: HTTP API 搜索 '{keyword}'...", file=sys.stderr)
    result, err = run_api(keyword, days)

    if result is not None:
        print(f"[reddit] 通道1 成功: {result['total']} 条", file=sys.stderr)
        print(json.dumps(result, ensure_ascii=False, indent=2))
        return
    else:
        print(f"[reddit] 通道1 失败，尝试降级...", file=sys.stderr)
        if err:
            # 只打印前 300 字符避免刷屏
            print(err[:300], file=sys.stderr)

    # ---- 通道 2：降级到浏览器方案 ----
    print(f"[reddit] 通道2: 浏览器搜索 '{keyword}'...", file=sys.stderr)
    result = run_browser(keyword, days)
    print(f"[reddit] 通道2 完成: {result.get('total', 0)} 条", file=sys.stderr)
    print(json.dumps(result, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
